### ISSUE TITLE
babel compiled code use wrong module name if the module names 'args'

### ISSUE THREAD

**Issue by xuhong** [babel/babel#6057]
try it link: 
http://babeljs.io/repl/#?babili=false&evaluate=false&lineWrap=true&presets=es2015%2Creact%2Cstage-0&targets=&browsers=&builtIns=false&debug=false&code_lz=JYWwDg9gTgLgBAJQKYEMDGMA0cDecDCE4EAdkifAL5wBmURcA5FKhowNwBQokscKUAOYBnWvRBMArjGAAbYQHpJUWQoEiOnTkgAeveABMkNFJNnw0slMNEBBMGDi6Y5A6MLEyFXJzhO90DAAKgCeYEhwALxMjFp-aESQXjAAIsAGALIQkhQAFACUPn5-MAAWwMIAdLr6oeFR_EJVgkgwuYwwYUiM2ADuwCQGEL2VshBoKDKklaUsNPlcfpRxcCyDSFAFRcUsMMokcCRmsux-vnDLlEA

### Input Code
see try it link upside

### Babel Configuration (.babelrc, package.json, cli command)
see try it link upside

### Expected Behavior
expected to call `_args2.default.get`

### Current Behavior
call `args.get`, which is `undefined`

### Your Environment
see try it link upside

**Comment by jridgewell** [comment]
This is an awesome bug. I'm pretty sure it takes:

1. A default import named `args`
2. A subclasses extending a superclass _without_ a constructor, or uses `...args` as a rest  param (and references it).
3. The imported `args` is used inside the class body

**Comment by jridgewell** [comment]
Well, this got out of my expertise quickly. I'm able to reduce this issue down to using `#insertBefore` to insert a for loop that declares a variable that's already block-scope declared in a higher scope. http://astexplorer.net/#/gist/319f21b8862c6b5777254c56a6bb9b03/e40094263a6b329085d30a564197cf7482015bf5

It's something to do with `Context`s. The `ForStatement` is requeued in the context of the `Program`, which causes it recrawl the bindings.

I'm nominating @loganfsmyth for it.

**Comment by loganfsmyth** [comment]
Bugs around `context` are such a pain because it is such a core piece of the current transform system and it's so hard to follow :(

**Comment by jridgewell** [comment]
What if we took away parent caching on the `Scope` entirely? We'd have to crawl `parentPath` until we find the real parent scope every time, but this bug would go away.

**Comment by jridgewell** [comment]
Ran into this again:

```js
const a = 'bar';
function foo(...a) {
  return a;
}
```
```json
{
  "plugins": ["transform-es2015-parameters"]
}
```
